src/hotspot/share/prims/jni.cpp

```diff
@@ -3836,6 +3838,9 @@ static jint attach_current_thread(JavaVM *vm, void **penv, void *_args, bool dae
     return JNI_ERR;
   }
 
+  // Please keep this inside the _attaching_via_jni section.
```

Comment:
Suggestion:
```
// Ensure the thread_id is set whilst still `_attaching_via_jni` so that it can be seen
// once the thread is included by the JFR thread iterator.
```
?

Reply:
Thanks David. I realized there is no need to move anything in jni.cpp, so I rolled them back. The change to the iterator will suffice.

Reply:
It will? Don't you still have the problem that the thread can be seen after ` thread->set_done_attaching_via_jni();` but before `post_thread_start_event(thread);` where the id will be set?

Reply:
Yes. The JFR_JVM_THREAD_ID(thread) is a lazy assignment of the thread id. It can be done by the thread itself, or the periodic task thread. The premise for a JavaThread is that there exist a threadObj.

Reply:
Now I get it. The zero thread id only arises when the `threadObj` is null and that is no longer possible when we skip JNI-attaching threads.